File: app/src/main/java/me/pwcong/jpstart/adapter/BasePagerAdapter.java
package me.pwcong.jpstart.adapter;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;

import java.util.List;

/**
 * Created by Pwcong on 2016/9/27.
 */

public abstract class BasePagerAdapter<T> extends FragmentPagerAdapter {

    private List<T> list;

    public BasePagerAdapter(FragmentManager fm, List<T> list) {
        super(fm);
        this.list = list;
    }

    @Override
    public Fragment getItem(int position) {
        return getFragment(list.get(position));
    }

    @Override
    public int getCount() {
        return list.size();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return getTitle(list.get(position));
    }

    protected abstract Fragment getFragment(T t);

    protected abstract CharSequence getTitle(T t);
}

File: app/src/main/java/com/github/pwcong/jpstart/adapter/BasePagerAdapter.kt
package com.github.pwcong.jpstart.adapter

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter

abstract class BasePagerAdapter<T>(fm: FragmentManager, private val list: List<T>) :
    FragmentPagerAdapter(fm) {
    override fun getItem(position: Int): Fragment {
        return getFragment(list[position])
    }

    override fun getCount(): Int {
        return list.size
    }

    override fun getPageTitle(position: Int): CharSequence? {
        return getTitle(list[position])
    }

    protected abstract fun getFragment(t: T): Fragment

    protected abstract fun getTitle(t: T): CharSequence?
}
